doc/build-overview.md

```diff
@@ -0,0 +1,251 @@
+<div class="hidden-warning"><a href="https://docs.haskellstack.org/"><img src="https://rawgit.com/commercialhaskell/stack/master/doc/img/hidden-warning.svg"></a></div>
+
+# Build Overview
+
+__NOTE__ This document should *not be considered accurate* until this
+note is removed.
+
+This is a work-in-progress document covering the build process used by Stack.
+Stack. It was started following the Pantry rewrite work in Stack (likely to
+land as Stack 2.0), and contains some significant changes/simplifications from
+how things used to work. This document will likely not fully be reflected in
+the behavior of Stack itself until late in the Stack 2.0 development cycle.
+
+## Terminology
+
+* Project package: anything listed in `packages` in stack.yaml
+* Dependency: anything listed in extra-deps or a snapshot
+* Target: package and/or component listed on the command line to be built. Can
+  be either project package or dependency. If none specified, automatically
+  targets all project packages
+* Immutable package: a package which comes from Hackage, an archive, or a
+  repository. In contrast to...
+* Mutable package: a package which comes from a local file path. The contents
+  of such a package are assumed to mutate over time.
+* Snapshot database: a package database and set of executables for a given set
+  of _immutable_ packages. Only packages from immutable sources and which
+  depend exclusively on other immutable packages can be in this database.
+  *QUESTION* Would this better be called the _immutable database_ or _write only
+  database_?
+* Local database: a package database and set of executables for packages which
+  are either mutable or depend on such mutable packages. Importantly, packages
+  in this database can be unregister, replaced, etc, depending on what happens
+  with the source packages.
+
+Outdated terminology to be purged:
+
+* Wanted
+* Local
+* Snapshot package
+
+## Inputs
+
+Stack pays attention to the following inputs:
+
+* Current working directory, used for finding the default `stack.yaml` file and
+  resolving relative paths
+* The `STACK_YAML` environment variable
+* Command line arguments (CLI args), as will be referenced below
+
+Given these inputs, Stack attempts the following process when performing a build.
+
+## Find the `stack.yaml` file
+
+* Check for a `--stack-yaml` CLI arg, and use that
+* Check for a `STACK_YAML` env var
+* Look for a `stack.yaml` in this directory or ancestor directories
+* Fall back to the default global project
+
+This file is parsed to provide the following config values:
+
+* `resolver`
+* `compiler`
+* `packages`
+* `extra-deps`
+* `flags`
+* `ghc-options`
```

Comment:
I'd add a line specifying what's done if a `stack.yaml` file only has a subset of these, for completeness.

Reply:
I'll add defaulting comments